src/cargo/util/config/mod.rs

```diff
@@ -1057,6 +1059,103 @@ impl Config {
         }
     }
 
+    /// Checks if the given path is owned by a different user.
+    fn check_safe_dir(&self, path: &Path, safe_directories: &HashSet<PathBuf>) -> CargoResult<()> {
+        if !self.safe_directories_enabled() {
+            return Ok(());
+        }
+        paths::validate_ownership(path, safe_directories).map_err(|e| {
+            match e.downcast_ref::<paths::OwnershipError>() {
+                Some(e) => {
+                    let mut to_add = e.path.parent().unwrap();
+                    if to_add.file_name().and_then(|f| f.to_str()) == Some(".cargo") {
+                        to_add = to_add.parent().unwrap();
+                    }
+                    ownership_error(e, "config files", to_add, self)
+                }
+                None => e,
+            }
+        })
+    }
+
+    /// Returns whether or not the nightly-only safe.directories behavior is enabled.
+    pub fn safe_directories_enabled(&self) -> bool {
+        // Because the config is loaded before the CLI options are available
+        // (primarily to handle aliases), this can't be gated on a `-Z` flag.
+        // So for now, this is only enabled via an environment variable. This
+        // has some downsides (such as not supporting the "allow" list), but
+        // should be fine for a one-off exception.
+        self.env
+            .get("CARGO_UNSTABLE_SAFE_DIRECTORIES")
+            .map(|s| s.as_str())
+            == Some("true")
+            && self.nightly_features_allowed
+    }
+
+    /// Returns the `safe.directories` config setting.
+    pub fn safe_directories(&self) -> CargoResult<&HashSet<PathBuf>> {
+        self.safe_directories.try_borrow_with(|| {
+            if !self.safe_directories_enabled() {
+                return Ok(HashSet::new());
+            }
+            let mut safe_directories: HashSet<PathBuf> = self
+                .get_list("safe.directories")?
+                .map(|dirs| {
+                    dirs.val
+                        .iter()
+                        .map(|(s, def)| def.root(self).join(s))
+                        .collect()
+                })
+                .unwrap_or_default();
+            self.extend_safe_directories_env(&mut safe_directories);
+            Ok(safe_directories)
+        })
+    }
+
+    fn extend_safe_directories_env(&self, safe_directories: &mut HashSet<PathBuf>) {
+        // Note: Unlike other Cargo environment variables, this does not
+        // assume paths relative to the current directory (only absolute paths
+        // are supported). This is intended as an extra layer of safety.
```

Comment:
- I didn't see this in the RFC
- For my use of git's equivalent, I did use a `.` though I did it through `-c` rather than an env variable
- We are only enforcing this contract by the fact that the a `.` won't match anything, correct?  Should we more explicitly tell the user, either via a warning or error, that their safe directory setting will do nothing?